coap/Kconfig

```diff
@@ -69,4 +73,37 @@ menu "CoAP Configuration"
         default 7 if COAP_LOG_DEBUG
         default 9 if COAP_LOG_MBEDTLS
 
+    config COAP_TCP_SUPPORT
+        bool "Enable TCP within CoAP"
+        default y
+        help
+            Enable TCP functionality for CoAP. This is required if TLS sessions
+            are to be used.
+
+            If this option is disabled, redundent CoAP TCP code is removed.
+
+    config COAP_CLIENT_SUPPORT
+        bool "Enable Client functionality within CoAP"
+        default n
```

Comment:
I'm sorry if I am misunderstanding this part, but is there any downside setting this and `COAP_SERVER_SUPPORT` to `y` by default?

Since IDF compiles the code with `-ffunction-sections -Wl,--gc-sections`, we expect that most of the unused code will be removed if it is not explicitly referenced in the application, anyway.

On the other hand, if an existing project updates its libcoap version, it won't need to modify sdkconfig file setting these options to 'y'. (Otherwise this could be seen as a breaking change.)

Reply:
Good question, which I had to think about.

If both COAP_CLIENT_SUPPORT and COAP_SERVER_SUPPORT are not set, then libcoap builds as if both were set (for backwards compatibility).  So any existing project code will compile the new libcoap and work 'as is'.

I have got some pending work ready for a PR for the general esp-idf which updates sdkconfig.defaults (for example)
````
diff --git a/examples/protocols/coap_server/sdkconfig.defaults b/examples/protocols/coap_server/sdkconfig.defaults
index 308ca1c49a..8bc64e0867 100644
--- a/examples/protocols/coap_server/sdkconfig.defaults
+++ b/examples/protocols/coap_server/sdkconfig.defaults
@@ -2,3 +2,4 @@ CONFIG_MBEDTLS_SSL_PROTO_DTLS=y
 CONFIG_MBEDTLS_PSK_MODES=y
 CONFIG_MBEDTLS_KEY_EXCHANGE_PSK=y
 CONFIG_LWIP_NETBUF_RECVINFO=y
+CONFIG_COAP_SERVER_SUPPORT=y
````
to set a sensible default.

There are a fair number of functions that have both client and server code in them, so there are code savings there.  I agree that dedicated server / client functions will get removed, but there may be parts of the library code that call these unused functions.

Building with both client and server true, for the examples coap_server, I get
````
Per-archive contributions to ELF file:
            Archive File DRAM .data .rtc.data DRAM .bss IRAM0 .text & 0.vectors ram_st_total Flash .text & .rodata & .appdesc flash_total
         libmbedcrypto.a         40         0        68          30           0          138       88937     63178          0      152185
               liblwip.a         12         0      3706           0           0         3718       97813     15715          0      113540
            libmbedtls.a         84         0       380           0           0          464       71808     32767          0      104659
    libespressif__coap.a        284         0     12690           0           0        12974       69108     17542          0       86934

````
With server mode only, I get
````
Per-archive contributions to ELF file:
            Archive File DRAM .data .rtc.data DRAM .bss IRAM0 .text & 0.vectors ram_st_total Flash .text & .rodata & .appdesc flash_total
         libmbedcrypto.a         40         0        68          30           0          138       88937     63178          0      152185
               liblwip.a         12         0      3706           0           0         3718       93915     15519          0      109446
            libmbedtls.a         84         0       380           0           0          464       71349     32767          0      104200
    libespressif__coap.a        276         0     12678           0           0        12954       60736     16022          0       77034
````
where there are also marginal savings in  liblwip.a and libmbedtls.a.

Reply:
Thank you for the explanation! All the points make sense to me.